File: testprotobuf.py
from contacts_pb2 import Contact, PhoneNumber

contact = Contact(
    index=1,
    first_name="John",
    last_name="Smith",
)

contact.phone_numbers.append(
    PhoneNumber(
        phone_type="home",
        number="123-456-7890",
    )
)
contact.phone_numbers.append(
    PhoneNumber(
        phone_type="work",
        number="012-345-6789",
    )
)

print(contact)
data = contact.SerializeToString() # serializes to bytes object
# TODO: check that data equals some byte string
recovered = Contact()
recovered.ParseFromString(data)
print(recovered)

File: testprotobuf.js
const protobuf = require("protobufjs");

async function main() {
    const root = await protobuf.load("contacts.proto");
    const Contact = root.lookupType("contacts.Contact");
    const PhoneNumber = root.lookupType("contacts.PhoneNumber");

    const home = PhoneNumber.create({
        phoneType: "home",
        number: "123-456-7890"
    });
    const work = PhoneNumber.create({
        phoneType: "work",
        number: "012-345-6789"
    });

    const contact = Contact.create({
        index: 1,
        firstName: "John",
        lastName: "Smith",
        phoneNumbers: [home, work]
    });

    console.log(contact);
    const data = Contact.encode(contact).finish(); // serializes to Uint8Array
    console.log(data.length === 59);
    const recovered = Contact.decode(data);
    console.log(recovered);
}

main();
